src/datatypes/src/vectors/decimal.rs

```diff
@@ -0,0 +1,465 @@
+    /// And check precision and scale if compatible.
+    pub fn with_precision_and_scale(self, precision: u8, scale: i8) -> Result<Self> {
+        let array = self
+            .array
+            .with_precision_and_scale(precision, scale)
+            .context(InvalidArgumentsSnafu {})?;
+        Ok(Self { array })
+    }
+
+    pub fn null_if_overflow_precision(&self, precision: u8) -> Self {
+        Self {
+            array: self.array.null_if_overflow_precision(precision),
+        }
+    }
+
+    pub fn validate_decimal_precision(&self, precision: u8) -> Result<()> {
+        self.array
+            .validate_decimal_precision(precision)
+            .context(InvalidArgumentsSnafu {})
+    }
+
+    /// Return decimal value as string
+    pub fn value_as_string(&self, idx: usize) -> String {
+        self.array.value_as_string(idx)
+    }
+
+    pub fn precision(&self) -> u8 {
+        self.array.precision()
+    }
+
+    pub fn scale(&self) -> i8 {
+        self.array.scale()
+    }
+
+    pub(crate) fn as_arrow(&self) -> &dyn Array {
+        &self.array
+    }
+}
+
+impl Vector for Decimal128Vector {
+    fn data_type(&self) -> ConcreteDataType {
+        if let ArrowDataType::Decimal128(p, s) = self.array.data_type() {
+            ConcreteDataType::decimal128_datatype(*p, *s)
+        } else {
+            ConcreteDataType::decimal128_default_datatype()
+        }
+    }
+
+    fn vector_type_name(&self) -> String {
+        "Decimal128Vector".to_string()
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn len(&self) -> usize {
+        self.array.len()
+    }
+
+    fn to_arrow_array(&self) -> ArrayRef {
+        let data = self.array.to_data();
+        Arc::new(Decimal128Array::from(data))
+    }
+
+    fn to_boxed_arrow_array(&self) -> Box<dyn Array> {
+        let data = self.array.to_data();
+        Box::new(Decimal128Array::from(data))
+    }
+
+    fn validity(&self) -> Validity {
+        vectors::impl_validity_for_vector!(self.array)
+    }
+
+    fn memory_size(&self) -> usize {
+        self.array.get_buffer_memory_size()
+    }
+
+    fn null_count(&self) -> usize {
+        self.array.null_count()
+    }
+
+    fn is_null(&self, row: usize) -> bool {
+        self.array.is_null(row)
+    }
+
+    fn slice(&self, offset: usize, length: usize) -> VectorRef {
+        let data = self.array.to_data().slice(offset, length);
+        Arc::new(Self::from_array_data(data))
+    }
+
+    fn get(&self, index: usize) -> Value {
+        if !self.array.is_valid(index) {
+            return Value::Null;
+        }
+
+        match self.array.data_type() {
+            ArrowDataType::Decimal128(precision, scale) => {
+                // Safety: The index have been checked by `is_valid()`.
+                let value = unsafe { self.array.value_unchecked(index) };
+                Value::Decimal128(Decimal128::new_unchecked(value, *precision, *scale))
```

Comment:
`_unchecked` is usually for unsafe functions. I'd suggest replacing `new_unchecked` with `new` and asserting the range of precision and scale in it.

If you really want to use a function that doesn't do any validation, you should provide an `unsafe` version `new_unchecked()`. But can we guarantee safety while using `new_unchecked()`?